pageserver/src/walredo.rs

```diff
@@ -202,9 +202,21 @@ impl PostgresRedoManager {
     /// rely on our owner calling this function periodically in its own housekeeping
     /// loops.
     pub(crate) fn maybe_quiesce(&self, idle_timeout: Duration) {
-        if let Ok(g) = self.last_redo_at.try_lock() {
-            if let Some(last_redo_at) = *g {
-                if last_redo_at.elapsed() >= idle_timeout {
+        if let Ok(g) = self.last_successful_redo_at.try_lock() {
+            if let Some(last_successful_redo_at) = *g {
+                // Kill the walredo process if
+                // - it has been unused for `idle_timeout`
+                // - it has been used, but, without success.
+                // The former is just good housekeeping.
+                // The latter adds robustness for the case where something is wrong
+                // with the walredo process.
+                //
+                // Note that we don't want to kill the process immediately on each redo failure.
+                // The reason is that the redo failure could be caused by corrupted or malicious data.
+                // We don't want to get into a kill-respawn loop in that case.
+                // So, we piggy-back on the quiescing mechanism,
+                // resulting in a max kill-respawn frequency of `1/idle_timeout`.
```

Comment:
I don't understand this either. Wouldn't this require writing `Some(Instant::now())` on launching a new process?

Reply:
Right now:
- if we never have successful walredo: never killed by this (probably improbable)
- if we have 1 successful walredo and fail always after, this will always kill the process after timeout has elapsed